### PR TITLE
add Validator.test and clone validationmethod

### PR DESCRIPTION
**Pull request by zhangxiaoli73**
## What changes were proposed in this pull request?
1. clone validationmethod in optimizer 
2. add unit test for Validator.test

## How was this patch tested?

add unit test

## Related links or issues (optional)
fixed https://github.com/intel-analytics/BigDL/issues/523

### CODE REVIEW

File: dl/src/main/scala/com/intel/analytics/bigdl/optim/ValidationMethod.scala [hunk 1]
```diff
@@ -205,9 +206,10 @@ class LossResult(private var loss: Float, private var count: Int)
   }
 }
 
-class Loss[@specialized(Float, Double)T: ClassTag]()
+class Loss[@specialized(Float, Double)T: ClassTag](
+ criterion: Criterion[T])
```
Comment (jason-dai) [inline dl/src/main/scala/com/intel/analytics/bigdl/optim/ValidationMethod.scala]: Add a default value (ClassNLLCriterion) for criterion

File: dl/src/main/scala/com/intel/analytics/bigdl/python/api/PythonBigDL.scala [hunk 2]
```diff
@@ -72,7 +72,7 @@ class PythonBigDL[T: ClassTag](implicit ev: TensorNumeric[T]) extends Serializab
     vMethods.toArray.map{case m: String => m.toLowerCase()}.map {
       case "top1" => new Top1Accuracy[T]()
       case "top5" => new Top5Accuracy[T]()
-      case "loss" => new Loss[T]()
+      case "loss" => new Loss[T](new ClassNLLCriterion[T]())
```
Comment (jason-dai) [inline dl/src/main/scala/com/intel/analytics/bigdl/python/api/PythonBigDL.scala]: We should allow the python user to specify what criterion to use for testing "loss". @zhichao-li

File: dl/src/main/scala/com/intel/analytics/bigdl/optim/DistriOptimizer.scala [hunk 3]
```diff
@@ -452,8 +454,9 @@ object DistriOptimizer {
               val input = batch.data.narrow(1, offset + 1, length)
               val target = batch.labels.narrow(1, offset + 1, length)
               val output = workingModels(b).forward(input)
-              vMethods.map(validation => {
-                validation(output, target, criterion(b))
+              val validatMethods = method.clone()
```
Comment (yiheng) [inline dl/src/main/scala/com/intel/analytics/bigdl/optim/DistriOptimizer.scala]: why clone in each batch?

File: dl/src/main/scala/com/intel/analytics/bigdl/optim/LocalOptimizer.scala [hunk 4]
```diff
@@ -217,8 +217,9 @@ class LocalOptimizer[T: ClassTag] private[optim](
             val input = batch.data.narrow(1, offset + 1, length)
             val target = batch.labels.narrow(1, offset + 1, length)
             val output = workingModels(b).forward(input)
-            vMethods.map(validation => {
-              validation(output, target, workingCriterion(b))
+            val validatMethods = vMethods.clone()
```
Comment (yiheng) [inline dl/src/main/scala/com/intel/analytics/bigdl/optim/LocalOptimizer.scala]: don't clone each time. use a buffer of vMehtods instead.
vMethods: Array[vMethod]
vMethods.length == parallelism
replace vMethods.clone() with vMethods(b)

File: dl/src/main/scala/com/intel/analytics/bigdl/optim/ValidationMethod.scala [hunk 5]
```diff
@@ -124,10 +125,11 @@ class Top1Accuracy[T] extends ValidationMethod[T] {
 }
 
 class Top5Accuracy[T] extends ValidationMethod[T] {
-  override def apply(output: Activity, target: Activity, criterion: Criterion[T] = null):
+  override def apply(output: Activity, target: Activity):
   AccuracyResult = {
     val _output = output.asInstanceOf[Tensor[T]]
     val _target = target.asInstanceOf[Tensor[T]]
+    _target.squeeze()
```
Comment (yiheng) [inline dl/src/main/scala/com/intel/analytics/bigdl/optim/ValidationMethod.scala]: should not modify the original ground truth data. You may need to add a new squeeze() method which return a new tensor point to the same storage with a different view.

File: dl/src/main/scala/com/intel/analytics/bigdl/optim/DistriOptimizer.scala [hunk 6]
```diff
@@ -425,18 +426,20 @@ object DistriOptimizer {
...
       return
     }
-    val vMethods = validationMethods.get
     val validateRDD = validationDataSet.get.toDistributed().data(train = false)
     logger.info(s"[Wall Clock ${wallClockTime / 1e9}s] Validate model...")
     val _subModelNumber = Engine.getEngineType match {
       case MklBlas => coresPerNode
       case _ => throw new IllegalArgumentException
     }
+    val vMethods = (1 to _subModelNumber).map(i => {
```
Comment (yiheng) [inline dl/src/main/scala/com/intel/analytics/bigdl/optim/DistriOptimizer.scala]: What's this?

File: dl/src/main/scala/com/intel/analytics/bigdl/tensor/DenseTensor.scala [hunk 7]
```diff
@@ -59,6 +59,11 @@ private[tensor] class DenseTensor[@specialized(Float, Double) T: ClassTag](
 
   override def squeeze(dim: Int): Tensor[T] = DenseTensor.squeeze(this, dim - 1)
 
+  override def squeezeNewTensor(): Tensor[T] = {
+    val result = Tensor(this.storage()).resizeAs(this)
```
Comment (yiheng) [inline dl/src/main/scala/com/intel/analytics/bigdl/tensor/DenseTensor.scala]: I think you should not call Tensor() in a DenseTensor.

Comment (zhangxiaoli73) [inline dl/src/main/scala/com/intel/analytics/bigdl/tensor/DenseTensor.scala]: yeah, changed

File: dl/src/main/scala/com/intel/analytics/bigdl/optim/DistriOptimizer.scala (hunk 6 above)
Comment (zhangxiaoli73) [inline dl/src/main/scala/com/intel/analytics/bigdl/optim/DistriOptimizer.scala]: ok, changed

**Comment by zhangxiaoli73** [comment]
jenkins pass

**Pull request merged**